### ISSUE TITLE
Implement DHT in reasonml over webrtc.

### ISSUE THREAD

**Issue by Schmavery** [Schmavery/raccoon#2]
Chord is a DHT (distributed hash table) scheme.  This is how we want to share data throughout the network.  There could be an argument for using some sort of gossip protocol instead, but from what I can tell, this seems to scale better. I welcome lots of discussion here.

There is some previous work that has been done in this area that should be considered:
https://github.com/diasdavid/webrtc-explorer
https://www.youtube.com/watch?v=fNQGGGE__zI&feature=youtu.be&t=13m33s
TL;DR: He had to tweak the finger table to support the limitations of webrtc.

Information about Chord:
https://en.wikipedia.org/wiki/Chord_(peer-to-peer)
Original paper: https://pdos.csail.mit.edu/papers/chord:sigcomm01/chord_sigcomm.pdf

More reading:
https://en.wikipedia.org/wiki/Sybil_attack - I think this attack only applies to systems with reputation systems, which things like facebook don't have, so we should be safe even though Chord is vulnerable.

**Comment by Schmavery** [comment]
Followup:
It seems like Kademlia might be a better option for this.  It has many more practical implementations than Chord (which seems to be more of an academic dht algorithm).  I read the [paper](https://pdos.csail.mit.edu/~petar/papers/maymounkov-kademlia-lncs.pdf) and was impressed with a couple details that they added that should optimize the performance (even if it doesn't improve the asymptotic runtime complexity).

I'm going ahead with attempting to get some implementation of this working using the webrtc bindings I wrote for bucklescript.  

Again, the limitations of webrtc will force us to make several changes to the algorithm.  Namely, all the entries in the k-buckets will need to be active webrtc connections rather than tuples of <id, udp port> because all connections need an intermediary to help facilitate the initial exchange.  There's a limit of around 200 active webrtc connections (and we have no indication of performance at that level).  This will result in a few necessary changes to the base protocol and we will see as we go what effect this will have on the system.
